### PR TITLE
[WIP] Quantization Refactor

### PR DESCRIPTION
**Pull request by TonyCongqianWang**
Improves quant config and removes hard coded values from the code.

The main goal is to improve readability and maintainability. As of currently there is a mismatch between training and inference (see #416 for more information). This patch intends to not only fix the mismatch but also to prevent mismatches in the future by inferring runtime correction factors directly from the quantization config instead of relying on manual changes across the codebase.

This allows arbitrary changes to the weight and activation quantization scales without having to make changes across the entire codebase.

As a quality of life improvement, serialize.py and quantize.py were refactored and the information density of the histogram callback was improved: The current histogram callback reports three values that are rather uninformative, since the model weights are already clipped through the callbacks. The logic was changed to use the native clipping of the model first and then clip only for type safety for serialization.

Allows more fine grained control over quant scheme (i.e. different weight scales per ls layer) and automatically calculates required adjustment factors to match inference instead of relying on hardcoding

Keeps quant scheme as is with TODO comments for possible further improve. In particular empirical observations show that slightly different quant scales probably preferred.

ADDED FUNCTIONALITY:
- fake quantization of activations 

~~Output scale is handled by https://github.com/TonyCongqianWang/Stockfish/blob/384175653285f68b67bd6a8d696380c40731e8c8/src/nnue/nnue_architecture.h#L135 which should yield slightly higher resolution. The main advantage however is better maintainability.~~

~~The maximum number of fully saturated inputs to i32 with i8*i8 inputs would be `133,144 / (600 * OutputScale) = 13` There is actually some risk of overflow even with the current setup, as even L3 has enough neurons and L1 has a skip connection. However given that L1=8k worked, this should be fine. But as a safety measure I will change the order of operation so that the maximum number of fully saturated neurons becomes `133,144 / (600 * OutputScale / WeightScale) = 887`~~

### CODE REVIEW

File: model/quantize.py [hunk 1]
```diff
@@ -13,82 +13,129 @@ class WeightClippingConfig(TypedDict):
...
+        self.min_threat_weight = -_i8.max / config.ft_quantized_one  # -127/256
+        self.max_threat_weight = _i8.max / config.ft_quantized_one  # 127/256
+
+        self._l0_correction_factor = config.ft_quantized_one ** 2 / config.inference_l0_division_factor / self.hidden_quantized_one
+        self._sqr_crelu_correction_factor = config.hidden_quantized_one / config.inference_sqr_crelu_division_factor
+        self._max_ft_activation = config.ft_quantized_max / config.ft_quantized_one
+        self._max_hidden_activation = config.hidden_quantized_max / config.hidden_quantized_one
+
+    @property
+    def l0_correction_factor(self):
```
Comment (xu-shawn) [inline model/quantize.py]: why use `@property` for these variables when rest are directly accessed?

File: model/model.py [hunk 2]
```diff
@@ -77,23 +78,59 @@ def forward(
...
+        l0_ = torch.clamp(l0_, 0.0, self.quantization.max_ft_activation)
+        if fake_quantize_acts:
+            # Fake quantization with STE
+            # NOTE if we use fake quantization in more spots,
+            # we should refactor this into its own class
+            # Inference uses bitshift which is equivalent to rounding down (floor).
+            act_scale = self.quantization.config.inference_l0_division_factor
+            l0_hard = ((l0_ * act_scale).floor() / act_scale).detach()
+            l0_soft = l0_.detach()
+            l0_ = l0_hard + (l0_ - l0_soft)
```
[the remark this answers is not in the document]
Comment (xu-shawn) [reply model/model.py]: can we use torch's own FakeQuantize here?

[the remark this answers is not in the document]
Comment (TonyCongqianWang) [reply model/model.py]: > can we use torch's own FakeQuantize here?

torchs own FakeQuantize works somewhat differently. We could eventually move there but it would require some iterations. If we keep good hygiene though, the switch at the end should be quite painless

File: model/modules/layer_stacks.py [hunk 3]
```diff
@@ -30,13 +31,15 @@ def __init__(self, count: int, config: LayerStacksConfig):
     def forward(self, x: torch.Tensor, ls_indices: torch.Tensor):
         l1c_ = self.l1(x, ls_indices)
         l1x_, l1x_out = l1c_.split(self.L2, dim=1)
-        # multiply sqr crelu result by (255/256) to match quantized version
+        # multiply sqr crelu result by scale correction to match quantized version
         l1x_ = torch.clamp(
-            torch.cat([torch.pow(l1x_, 2.0) * (255 / 256), l1x_], dim=1), 0.0, 1.0
+            torch.cat([torch.pow(l1x_, 2.0) * (self.quantization.sqr_crelu_correction_factor), l1x_], dim=1),
+            0.0,
+            self.quantization.max_hidden_activation,
```
[the remark this answers is not in the document]
Comment (TonyCongqianWang) [reply model/modules/layer_stacks.py]: Not a concern. We dont care about outdated .pts

File: model/quantize.py [hunk 4]
```diff
@@ -13,82 +13,129 @@ class WeightClippingConfig(TypedDict):
...
-        raise OverflowError(
-            f"values [{t_min}, {t_max}] do not fit in {dtype} "
-            f"range [{info.min}, {info.max}]"
-        )
-    return tensor.to(dtype)
-
+    rounded_value = value.round()
+    clamped_value = rounded_value.clamp(min_val, max_val)
+    num_clipped = (rounded_value != clamped_value).sum()
+    quantized_value = clamped_value.to(target_dtype)
```
[the remark this answers is not in the document]
Comment (TonyCongqianWang) [reply model/quantize.py:26]: Why?
This should be an extreme edge case and it is not silent. As you mentioned it is output in the histogramm

File: model/quantize.py (hunk 1 above)
Comment (TonyCongqianWang) [reply model/quantize.py]: Good question... at some point it wasn't. Ill change it

File: model/model.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (TonyCongqianWang) [reply model/model.py]: fixed

File: ftperm.py [hunk 5]
```diff
@@ -589,19 +560,19 @@ def eval_ft(model: NNUEModel, batch: data_loader.SparseBatchPtr, device_str: str
...
+        l0_, wpsqt, bpsqt = model.forward_ft(
             us,
             them,
             white_indices,
             white_values,
             black_indices,
             black_values,
             psqt_indices,
-            layer_stack_indices,
+            fake_quantize_acts=True,
```
[the remark this answers is not in the document]
Comment (TonyCongqianWang) [reply ftperm.py:571]: Yes there is some mistake. but not using fake_quantize isnt the solution

File: model/quantize.py (hunk 1 above)
Comment (TonyCongqianWang) [reply model/quantize.py]: fixed

File: model/model.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (xu-shawn) [reply model/model.py]: what do you mean by some iterations and what's the difference?

File: model/quantize.py [hunk 6]
```diff
@@ -13,82 +13,142 @@ class WeightClippingConfig(TypedDict):
...
+        self.max_hidden_weight = [hidden_q_max / scale for scale in self.weight_scale_hidden]
+        # Threat weights are treated separately. A bit hacky...
+        # Threat weights are quantized to int8 after scaling by ft_quantized_one
         _i8 = torch.iinfo(torch.int8)
-        self.min_threat_weight = -_i8.max / config.ft_quantized_one  # -127/255
-        self.max_threat_weight = _i8.max / config.ft_quantized_one  # 127/255
-        self.max_out_weight = (
-            config.hidden_quantized_one * self.hidden_quantized_one
-        ) / (self.nnue2score * self.weight_scale_out)
+        self.min_threat_weight = -_i8.max / config.ft_quantized_one  # -127/256
```
Comment (xu-shawn) [inline model/quantize.py:85]: isn't quantized_one 255?

**Comment by TonyCongqianWang** [comment]
Superseeded by #477
